### ISSUE TITLE
Application does not start anymore with only AbstractRoutingDataSource beans

### ISSUE THREAD

**Issue by tokuhirom** [spring-projects/spring-boot#18661]
`DataSourceHealthContributorAutoConfiguration` loads all `DataSource`s. And monitoring all data sources with an actuator.

It ignores `AbstractRoutingDataSource...` So it's a good implementation since the destination of `AbstractRoutingDataSource` may change.

https://github.com/spring-projects/spring-boot/blob/b43827d6264a3eb4e9558c6e1dd7f89e537caa96/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/jdbc/DataSourceHealthContributorAutoConfiguration.java#L79-L96

But `createContributor` throws an exception when the argument is empty.

https://github.com/spring-projects/spring-boot/blob/b43827d6264a3eb4e9558c6e1dd7f89e537caa96/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/health/AbstractCompositeHealthContributorConfiguration.java#L51-L57

As a result, if the module have only `AbstractRoutingDataSources`, boot application can't run.

**Comment by snicoll** [comment]
Thanks for the report. As far as I can see, we were checking the map isn't `null` in `2.1.x` and the check is a bit more complete in `2.2.x`. We need to figure out what to do if the map is empty.

**Comment by tokuhirom** [comment]
Here's a sample code to reproduce the issue.

https://github.com/tokuhirom/boot22-issue18661

**Comment by snicoll** [comment]
Here is the stacktrace

```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.actuate.health.HealthContributor]: Factory method 'dbHealthContributor' threw exception; nested exception is java.lang.IllegalArgumentException: Beans must not be empty
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.2.0.RELEASE.jar:5.2.0.RELEASE]
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:640) ~[spring-beans-5.2.0.RELEASE.jar:5.2.0.RELEASE]
        ... 105 common frames omitted
Caused by: java.lang.IllegalArgumentException: Beans must not be empty
        at org.springframework.util.Assert.notEmpty(Assert.java:549) ~[spring-core-5.2.0.RELEASE.jar:5.2.0.RELEASE]
        at org.springframework.boot.actuate.autoconfigure.health.AbstractCompositeHealthContributorConfiguration.createContributor(AbstractCompositeHealthContributorConfiguration.java:52) ~[spring-boot-actuator-autoconfigure-2.2.0.RELEASE.jar:2.2.0.RELEASE]
        at org.springframework.boot.actuate.autoconfigure.jdbc.DataSourceHealthContributorAutoConfiguration.dbHealthContributor(DataSourceHealthContributorAutoConfiguration.java:82) ~[spring-boot-actuator-autoconfigure-2.2.0.RELEASE.jar:2.2.0.RELEASE]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_202]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_202]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_202]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_202]
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.2.0.RELEASE.jar:5.2.0.RELEASE]
        ... 106 common frames omitted
```

**Comment by snicoll** [comment]
For the time being you can set `management.health.db.enabled=false` in your application configuration.

**Comment by snicoll** [comment]
If have a fix for a quite involved condition as I couldn't figure out how to bypass registration. I'd like a review from someone on the team.

**Comment by snicoll** [comment]
We discussed this one at the team meeting and I realized I didn't ask or check what was happening with such setup in `2.1.x`.

Here's the output for the sample above and `2.1.9.RELEASE`:

```
{
  "status": "UP",
  "details": {
    "db": {
      "status": "UNKNOWN"
    },
    "diskSpace": {
      "status": "UP",
      "details": {
        "total": 499963170816,
        "free": 163580035072,
        "threshold": 10485760
      }
    }
  }
}
```

If we want this to be consistent, we could return a health indicator that returns `UNKNOWN` when no datasources are available. Perhaps we could use the opportunity to actually improve the situation?
